subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractContextualMultiVersionSpecRunner.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.gradle.api.specs.Spec;
+import org.gradle.util.CollectionUtils;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.collect.Iterators.getLast;
+
+public abstract class AbstractContextualMultiVersionSpecRunner<T extends AbstractContextualMultiVersionSpecRunner.VersionedTool> extends AbstractMultiTestRunner {
+    public static final String VERSIONS_SYSPROP_NAME = "org.gradle.integtest.versions";
+
+    protected abstract Collection<T> getAllVersions();
```

Comment:
Is an implied contract of this method that it must not return an empty collection? Currently, if it returns an empty collection, both `getFirstAvailable` and `getLastAvailable` will return null. Guava is null hostile so `Sets.newHashSet` will throw an exception. Is this intended?